osu.Game/Utils/ModUtils.cs

```diff
@@ -0,0 +1,151 @@
+                        return false;
+
+                    incompatibleTypes.Add(t);
+                }
+
+                // Add the new mod types, checking whether any match the incompatible types.
+                foreach (var t in mod.GetType().EnumerateBaseTypes())
+                {
+                    if (incomingTypes.Contains(t))
+                        return false;
+
+                    incomingTypes.Add(t);
+                }
+            }
+
+            return true;
+        }
+
+        /// <summary>
+        /// Checks that all <see cref="Mod"/>s in a combination appear within a set of allowed types.
+        /// </summary>
+        /// <remarks>
+        /// The set of allowed types must contain exact <see cref="Mod"/> types for the respective <see cref="Mod"/>s to be allowed.
+        /// </remarks>
+        /// <param name="combination">The <see cref="Mod"/> combination to check.</param>
+        /// <param name="allowedTypes">The set of allowed <see cref="Mod"/> types.</param>
+        /// <returns>Whether all <see cref="Mod"/>s in the combination are allowed.</returns>
+        public static bool CheckAllowed(IEnumerable<Mod> combination, IEnumerable<Type> allowedTypes)
+        {
+            var allowedSet = new HashSet<Type>(allowedTypes);
+
+            return combination.SelectMany(FlattenMod)
+                              .All(m => allowedSet.Contains(m.GetType()));
+        }
+
+        /// <summary>
+        /// Check the provided combination of mods are valid for a local gameplay session.
+        /// </summary>
+        /// <param name="mods">The mods to check.</param>
+        /// <param name="invalidMods">Invalid mods, if any where found. Can be null if all mods were valid.</param>
+        /// <returns>Whether the input mods were all valid. If false, <paramref name="invalidMods"/> will contain all invalid entries.</returns>
+        public static bool CheckValidForGameplay(IEnumerable<Mod> mods, out Mod[]? invalidMods)
+        {
+            mods = mods.ToArray();
+
+            List<Mod>? foundInvalid = null;
+
+            void addInvalid(Mod mod)
+            {
+                foundInvalid ??= new List<Mod>();
+                foundInvalid.Add(mod);
+            }
+
+            foreach (var mod in mods)
+            {
+                bool valid = mod.Type != ModType.System
+                             && mod.HasImplementation
+                             && !(mod is MultiMod);
+
+                if (!valid)
+                {
+                    // if this mod was found as invalid, we can exclude it before potentially excluding more incompatible types.
+                    addInvalid(mod);
+                    continue;
+                }
+
+                foreach (var type in mod.IncompatibleMods)
+                {
+                    foreach (var invalid in mods.Where(m => type.IsInstanceOfType(m)))
+                        addInvalid(invalid);
```

Comment:
This feels sort of ehh, because it partially duplicates `CheckCompatibleSet()`, except it's written in a slightly different way... Would maybe be nice to have the option to have that method spit out the subset of incompatible mods? Not 100% sure.

Reply:
I'll need to read over the one again; it seemed like it was not going to be usable for my purpose. Would definitely be good to combine if we can.